language-support/ts/codegen/src/TsCodeGenMain.hs

```diff
@@ -107,31 +111,75 @@ genModule curPkgId mod
+    serDefs = defDataTypes mod
+
+defDataTypes :: Module -> [DefDataType]
+defDataTypes mod = filter (getIsSerializable . dataSerializable) (NM.toList (moduleDataTypes mod))
 
-genDataDef :: PackageId -> ModuleName -> NM.NameMap Template -> DefDataType -> (([T.Text], [T.Text]), Set.Set ModuleRef)
-genDataDef curPkgId curModName tpls def = case unTypeConName (dataTypeCon def) of
+genDataDef :: PackageId -> Module -> NM.NameMap Template -> DefDataType -> (([T.Text], [T.Text]), Set.Set ModuleRef)
+genDataDef curPkgId mod tpls def = case unTypeConName (dataTypeCon def) of
     [] -> error "IMPOSSIBLE: empty type constructor name"
     _: _: _: _ -> error "IMPOSSIBLE: multi-part type constructor of more than two names"
-    [conName] -> genDefDataType curPkgId conName curModName tpls def
-    [c1, c2] -> ((makeNamespace $ map ("  " <>) typs, makeNamespace $ map ("  " <>) sers), refs)
+
+    [conName] -> genDefDataType curPkgId conName mod tpls def
+    [c1, c2] -> ((makeNamespace $ map ("  " <>) typs, []), refs)
       where
-        ((typs, sers), refs) = genDefDataType curPkgId c2 curModName tpls def
-        ns = c1 <> "_NS"
+        ((typs, _), refs) = genDefDataType curPkgId c2 mod tpls def
+        ns = c1
         makeNamespace stuff =
           [ "// eslint-disable-next-line @typescript-eslint/no-namespace"
-          , "export namespace " <> ns <> " {"] ++ stuff ++ ["} //namespace " <> ns] ++ [""]
+          , "export namespace " <> ns <> " {"] ++ stuff ++ ["} //namespace " <> ns]
 
-genDefDataType :: PackageId -> T.Text -> ModuleName -> NM.NameMap Template -> DefDataType -> (([T.Text], [T.Text]), Set.Set ModuleRef)
-genDefDataType curPkgId conName curModName tpls def =
+genDefDataType :: PackageId -> T.Text -> Module -> NM.NameMap Template -> DefDataType -> (([T.Text], [T.Text]), Set.Set ModuleRef)
+genDefDataType curPkgId conName mod tpls def =
     case dataCons def of
         DataVariant bs ->
           let
             (typs, sers) = unzip $ map genBranch bs
-            typeDesc = [""] ++ typs
-            serDesc  = ["() => jtv.oneOf<" <> conName <> typeParams <> ">("] ++ sers ++ [")"]
-          in
-          ((makeType typeDesc, makeSer serDesc), Set.unions $ map (Set.setOf typeModuleRef . snd) bs)
+            typeDesc = makeType ([""] ++ typs)
+            typ = conName <> typeParams -- Type of the variant.
+            assocDefs = assocDefDataTypes mod conName -- Associated types.
+            serDesc =
+              if not $ null paramNames -- Polymorphic type.
+              then -- Companion function.
+                let
+                  -- Any associated serializers.
+                  assocSers = map
+                    (\(n, d) -> serFromDef id n curPkgId mod tpls d)
+                    assocDefs
+                  -- The variant deserializer.
+                  function = (makeSer ( ["() => jtv.oneOf<" <> typ <> ">("] ++ sers ++ [")"]) & onLast (<> ";"));
```

Comment:
```suggestion
                  function = onLast (<> ";")) (makeSer (["() => jtv.oneOf<" <> typ <> ">("] ++ sers ++ [")"])
```
If you need parentheses anyway, why bother with the `&`?

Reply:
Matter of taste. I did it that way because it reads nicely putting the ";" at the end as it comes out in the code. Do you feel strongly that it should change?

Reply:
IMO, not using operators that are not really necessary makes the code more accessible for future Haskell newbies in our team.

Reply:
That is a reasonable argument against. It was cute but I'll drop it's use here.